tokio-util/src/task/task_tracker.rs

```diff
@@ -0,0 +1,660 @@
+//! Types related to the [`TaskTracker`] collection.
+//!
+//! See the documentation of [`TaskTracker`] for more information.
+
+use pin_project_lite::pin_project;
+use std::fmt;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::sync::{futures::Notified, Notify};
+
+#[cfg(feature = "rt")]
+use tokio::{
+    runtime::Handle,
+    task::{JoinHandle, LocalSet},
+};
+
+/// A task tracker used for waiting until tasks exit.
+///
+/// This is usually used together with [`CancellationToken`] to implement [graceful shutdown]. The
+/// `CancellationToken` is used to signal to tasks that they should shut down, and the
+/// `TaskTracker` is used to wait for them to finish shutting down.
+///
+/// The `TaskTracker` will also keep track of a `closed` boolean. This is used to handle the case
+/// where the `TaskTracker` is empty, but we don't want to shut down yet. This means that the
+/// [`wait`] method will wait until *both* of the following happen at the same time:
+///
+///  * The `TaskTracker` must be closed using the [`close`] method.
+///  * The `TaskTracker` must be empty, that is, all tasks that it is tracking must have exited.
+///
+/// When a call to [`wait`] returns, it is guaranteed that all tracked tasks have exited and that
+/// the destructor of the future has finished running. However, there might be a short amount of
+/// time where [`JoinHandle::is_finished`] returns false.
+///
+/// # Comparison to `JoinSet`
+///
+/// The main Tokio crate has a similar collection known as [`JoinSet`]. The `JoinSet` type has a
+/// lot more features than `TaskTracker`, so you should only use a `TaskTracker` when you need one
+/// of its unique features:
+///
+///  1. When tasks exit, a `TaskTracker` will allow the task to immediately free its memory.
+///  2. By not closing the `TaskTracker`, you can prevent [`wait`] from returning even if the
+///     `TaskTracker` is empty.
+///  3. A `TaskTracker` does not require mutable access to insert tasks.
+///  4. A `TaskTracker` can be cloned to share it with many tasks.
+///
+/// The first point is the most important one. With a [`JoinSet`], it keeps track of the return
+/// value of every task you insert into it. This means that if you keep inserting tasks and never
+/// call [`join_next`], then their return values will keep building up and take up memory, even
+/// if most of the tasks have already exited. This can cause you to run out of memory. With a
+/// `TaskTracker`, this does not happen. Once tasks exit, they are immediately removed from the
+/// `TaskTracker`.
```

Comment:
```suggestion
/// The first point is the most important one. With a [`JoinSet`], it keeps track of the return
/// value of every inserted task. This means that if the caller keeps inserting tasks and never
/// calls [`join_next`], then their return values will keep building up and consuming memory, _even
/// if_ most of the tasks have already exited. This can cause the process to run out of memory.
/// With a `TaskTracker`, this does not happen. Once tasks exit, they are immediately removed from the `TaskTracker`.
```